### PR TITLE
refactor: netsim

### PR DESCRIPTION
**Pull request by Arqu**
More or less a complete redo of netsim.

- Maintains all backwards compatibility so far
- Cleaned up repo
- Cleaned up the code so it's manageable and other people can actually work on it too
- Added multicast support
- Added support for parallel execution on jobs where we don't strictly measure perf. Job time drops 2-4x in my experience so far.
- Fixed up old / disabled code (visualizations & sniffing)
- Much more extensible now

### CODE REVIEW

**Comment by Arqu** [comment]
@flub tagged you in for a quick look around if you like, given you're our resident python expert and shown some interest in this bit.
Also suggestions welcome.

This part 1 of 3.

File: netsim/parsing/netsim.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    stats = []
+    collect = False
+    for line in lines:
+        if line.startswith('[ ID]'):
+            collect = True
+            continue
+        if collect and 'datagram' not in line:
+            parts = [x for x in line.split() if x]
+            p = parts.index('sec')
+            transfer = float(parts[p+1]) * {'KBytes': 1024, 'MBytes': 1024 ** 2, 'GBytes': 1024 ** 3}.get(parts[p+2], 1)
```
Comment (flub) [inline netsim/parsing/netsim.py:61]: Maybe make this inline dict a constant?

File: netsim/sniffer/process.py [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+            elif tn['name'].startswith('n_'):
+                n['type'] = 'nat'
+                n['id'] = tn['name']
+            
+        node_list.append(n)
+    return packets, js_packet_data, node_list
+
+class NetsimViz():
+
+    HOVER_JS = """
```
Comment (flub) [inline netsim/sniffer/process.py:108]: Maybe write this in a `.js` file next to this and find it using `__file__`?

(Yes, using `__file__` for this is frowned upon but I don't think this stuff is a proper python distribution that can use the right way to load other resources.  And that's already a bit easier and doesn't force you into python packaging world)

File: netsim/sniffer/process.py [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+        def isatty(self):
+            return False
+    
+    sys.stdout = PseudoNonTTY(sys.stdout) # disable color output on packet data for JS
+    
+    viz = NetsimViz()
+    viz.run(args)
+    viz.export(args['output'])
+
+    sys.stdout = t_stdout
```
Comment (flub) [inline netsim/sniffer/process.py:402]: This should all be in a `try... except` block so that you restore `sys.stdout` whatever happens.

**Review by flub — COMMENTED** [review]
I've only given this a rather superficial review, not tried to understand the design and how things interact.  Let me know if you'd like a more thorough review around the architecture - but that's a bunch more work :)

**Review by flub — DISMISSED** [review]
Also, I don't mean to hold anything up so consider this approved whenever you want to :)

**Comment by Arqu** [comment]
> Also, I don't mean to hold anything up so consider this approved whenever you want to :)

Thank you. I mostly tagged you in to have someone else from the team be aware of it's existence (but I also knew you would give some input :) ) It's more/better than expected. 

A lot of the code base was cobbled together from different pieces and copy paste didn't do me a great service to keep things aligned. Starting with a formatter is a great 1st step.

**Pull request merged**